modules/renter/repairscanner.go

```diff
@@ -0,0 +1,333 @@
+	// and prevent the same work from being repeated after the next restart.
+	//
+	// TODO / NOTE: This process isn't going to make sense anymore once we
+	// switch to chunk-based saving.
+	if saveFile {
+		err := r.saveFile(f)
+		if err != nil {
+			r.log.Println("error while saving a file after pruning some contracts from it:", err)
+		}
+	}
+
+	// Iterate through the set of newUnfinishedChunks and remove any that are
+	// completed.
+	totalIncomplete := 0
+	for i := 0; i < len(newUnfinishedChunks); i++ {
+		if newUnfinishedChunks[i].piecesCompleted < newUnfinishedChunks[i].piecesNeeded {
+			newUnfinishedChunks[totalIncomplete] = newUnfinishedChunks[i]
+			newUnfinishedChunks[i].progress = float64(newUnfinishedChunks[i].piecesCompleted) / float64(newUnfinishedChunks[i].piecesNeeded)
+			totalIncomplete++
+		}
+	}
+	newUnfinishedChunks = newUnfinishedChunks[:totalIncomplete]
+	return newUnfinishedChunks
+}
+
+// managedBuildChunkHeap will iterate through all of the files in the renter and
+// construct a chunk heap.
+func (r *Renter) managedBuildChunkHeap(hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) *chunkHeap {
+	// Loop through the whole set of files to build the chunk heap.
+	var ch chunkHeap
+	id := r.mu.Lock()
+	for _, file := range r.files {
+		unfinishedChunks := r.buildUnfinishedChunks(file, hosts, fcidToHPK)
+		ch = append(ch, unfinishedChunks...)
+	}
+	r.mu.Unlock(id)
+
+	// Init the heap.
+	heap.Init(&ch)
+	return &ch
+}
+
+// managedInsertFileIntoChunkHeap will insert all of the chunks of a file into the
+// chunk heap.
+func (r *Renter) managedInsertFileIntoChunkHeap(f *file, ch *chunkHeap, hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) {
+	id := r.mu.Lock()
+	unfinishedChunks := r.buildUnfinishedChunks(f, hosts, fcidToHPK)
+	for i := 0; i < len(unfinishedChunks); i++ {
+		heap.Push(ch, unfinishedChunks)
+	}
+	r.mu.Unlock(id)
+}
+
+// managedPrepareNextChunk takes the next chunk from the chunk heap and prepares
+// it for upload. Preparation includes blocking until enough memory is
+// available, fetching the logical data for the chunk (either from the disk or
+// from the network), erasure coding the logical data into the physical data,
+// and then finally passing the work onto the workers.
+func (r *Renter) managedPrepareNextChunk(ch *chunkHeap, hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) {
+	// Grab the next chunk, loop until we have enough memory, update the amount
+	// of memory available, and then spin up a thread to asynchronously handle
+	// the rest of the chunk tasks.
+	memoryAvailable := r.managedMemoryAvailableGet()
+	nextChunk := ch.Pop().(*unfinishedChunk)
+	for nextChunk.memoryNeeded > memoryAvailable {
+		select {
+		case newFile := <-r.newUploads:
+			r.managedInsertFileIntoChunkHeap(newFile, ch, hosts, fcidToHPK)
+		case <-r.newMemory:
+		case <-r.tg.StopChan():
```

Comment:
don't you need to take some action after `<-r.newMemory`?

Reply:
ah yeah, need to update `memoryAvailable`